### PR TITLE
Improve Connect In Windows

### PR DESCRIPTION
**Pull request by AskBojesen**
In the Windows implementation, the ConnectToDeviceNativeAsync method did not work properly.
It was especially seen in the case when:

1. Connect
2. Get services
3. Disconnect
4. Connect
5. Get services

This has now been solved

### CODE REVIEW

**Comment by AskBojesen** [comment]
Hi @janusw , I found a bug in the Windows implementation I made recently.
Do you want be to create an issue also - or just this PR?
(It is not an issue in the released 3.0.0, since my Windows implementation came afterwards)

**Comment by janusw** [comment]
> Do you want me to create an issue also - or just this PR?

PR only is fine with me in this case, but maybe you can describe the problem a bit better? You explained the situation in which it occurs, but not what exactly happens.

**Comment by AskBojesen** [comment]
The problem wasbelow lines.... In Windows the BLE connection is established when asking for services.
Using the device.GetServicesAsync() which has a lot of unwanted sideeffects caused problems when reconnecting.
Eg. number of KnownServices got to high - setting BleImplementation.CacheModeGetServices - and then restoring was a hack I made to get something working - now I have a solution I understand, that also works better

`BluetoothCacheMode restoremode = BleImplementation.CacheModeGetServices;
BleImplementation.CacheModeGetServices = BluetoothCacheMode.Uncached;
var services = device.GetServicesAsync(cancellationToken).Result;
BleImplementation.CacheModeGetServices = restoremode;`

**Review by janusw — COMMENTED** [review]
Hm, getting the services seems like a rather indirect way of triggering a connection. Isn't there a more direct method for connecting to a device on Windows?

On all other platforms, connecting to a device and getting the services are two separate steps ...

**Comment by AskBojesen** [comment]
> Hm, getting the services seems like a rather indirect way of triggering a connection. Isn't there a more direct method for connecting to a device on Windows?
> 
> On all other platforms, connecting to a device and getting the services are two separate steps ...

Nope... And the previous implementation did the same...

In Wndows one get this object :

https://learn.microsoft.com/en-us/uwp/api/windows.devices.bluetooth.bluetoothledevice

You can see, that is no connect method...

The object is created by calling this method:

https://learn.microsoft.com/en-us/uwp/api/windows.devices.bluetooth.bluetoothledevice.frombluetoothaddressasync

We you can read:

_Creating a [BluetoothLEDevice](https://learn.microsoft.com/en-us/uwp/api/windows.devices.bluetooth.bluetoothledevice?view=winrt-22621) object by calling this method alone doesn't (necessarily) initiate a connection. To initiate a connection, set [GattSession.MaintainConnection](https://learn.microsoft.com/en-us/uwp/api/windows.devices.bluetooth.genericattributeprofile.gattsession.maintainconnection) to true, or call an uncached service discovery method on BluetoothLEDevice, or perform a read/write operation against the device._

**Comment by AskBojesen** [comment]
Well in the hindsight I would like to try out the hindsight _[GattSession.MaintainConnection](https://learn.microsoft.com/en-us/uwp/api/windows.devices.bluetooth.genericattributeprofile.gattsession.maintainconnection) = true_ 

I think it will be more clean - what I do not like about using the _NativeDevice.GetGattServicesAsync_ with the connect side effect is that I/We do not store the services we get.

Let me try out the _GattSession.MaintainConnection_ flag before we merge this PR

**Comment by AskBojesen** [comment]
@janusw : I tried out the different approach in Windows, but it gave various problems.
See the added comments in code
This PR is not changing the existing connect method - we could make an issue exploring this further...

I vote for merging this at it is now

**Comment by AskBojesen** [comment]
I managed to get the [connection working - inspired by the Android implementation.
Introducing ConnectInternal and DisconnectInternal in the device

**Comment by AskBojesen** [comment]
@janusw : If you review and approve this one today - then I will be extra gratefully - because it's my birthday today!

I guess I will go thorough the entire Windows implementation in the next 1,5 month - because we need this to work in our application at work. My point is - I am testing my own stuff with a skeptic mind - this must simply be robust and functional for our needs. We appreciate the plugin much so far - it is alive and functional. You and former contributors have made something that actually works. Details in the interface could be improved - but the essential parts focus on get things working.
We appreciate that.

**Comment by janusw** [comment]
> @janusw : If you review and approve this one today - then I will be extra gratefully - because it's my birthday today!

Happy birthday! 🎂 

(I'm not gonna jump through this hoop though 😉 - but I'll try have a careful look in the next days).

> I guess I will go through the entire Windows implementation in the next 1,5 month - because we need this to work in our application at work. My point is - I am testing my own stuff with a skeptic mind - this must simply be robust and functional for our needs.

Sounds good. I hope others will do some testing of the Windows implementation as well.

> We appreciate the plugin much so far - it is alive and functional. You and former contributors have made something that actually works. Details in the interface could be improved - but the essential parts focus on get things working. We appreciate that.

Thanks for the positive feedback :)

Most of that praise actually goes to other/former contributors. I am mostly trying to keep this project alive after the previous maintainer left, but I only authored a small part of the code (and basically none of the Windows implementation).

> Details in the interface could be improved

There is always room for improvement 😄. And we're always welcoming PRs which make this improvement happen. 😉

**Comment by janusw** [comment]
> @janusw: I managed to get the connection working - inspired by the Android implementation. Introducing ConnectInternal and DisconnectInternal in the device.
> 
> (I think this is more clean that calling get services without needing the services)
> 
> It all started by investigating the MTU size (or MaxPduSize as Windows calls it)
> 
> Now I believe we have the best solution for the Windows implementation

Yes, I agree that this is better and cleaner than before.


> sorry for the editing during this PR, but I guess that is what review is in fact for - you got me thinking more about the solution.

That is absolutely what a review is meant for: Either to point out a specific problem, or to get you thinking deeper and trigger further improvements this way. I'm glad my review was a good one in that sense :)

File: Source/Plugin.BLE/Windows/Adapter.cs
Comment (janusw) [inline Source/Plugin.BLE/Windows/Adapter.cs:87]: You're adding extra whitespace here ...

File: Source/Plugin.BLE/Windows/Adapter.cs
Comment (janusw) [inline Source/Plugin.BLE/Windows/Adapter.cs:152]: ... and here. Not a big deal, you might say. But if people later edit this file with an editor that automatically removes this, their commits will touch these lines, which can hinder blaming and git archeology.

Comment (janusw) [reply Source/Plugin.BLE/Windows/Adapter.cs:152]: Sorry for being pedantic here 🙈

Comment (AskBojesen) [reply Source/Plugin.BLE/Windows/Adapter.cs]: Its's ok - I have formatted the documents now

**Comment by AskBojesen** [comment]
@janusw : I think your changes to the master branch is causing build status problems here.

**Comment by janusw** [comment]
> @janusw : I think your changes to the master branch is causing build status problems here.

Huh, strange, the CI builds were working on master, not sure why they fail here now (on the Android targets, which you did not even touch). Maybe due to changes in the GHA runner images, will check later.

**Comment by janusw** [comment]
I just re-triggered the CI builds on master, in order to see if the still work, or if they fail now ...

**Comment by janusw** [comment]
> I just re-triggered the CI builds on master, in order to see if they still work, or if they fail now ...

Indeed the Windows build fails with the same error now as on your branch, so this is obviously not your fault (and not your problem). I will look into it ...

**Comment by janusw** [comment]
Fixed the CI on master and rebased your branch. I hope it should succeed now ...

**Review by janusw — APPROVED** [review]
The CI builds succeed now. This PR is ok with me.

**Comment by janusw** [comment]
Merged. Thanks for the contribution!

**Pull request merged**
